modules/dns-helper/dns_helper_test.go

```diff
@@ -0,0 +1,289 @@
+package dns_helper
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+var publicDomainNameservers = []string{
+	"ns-1499.awsdns-59.org",
+	"ns-190.awsdns-23.com",
+	"ns-1989.awsdns-56.co.uk",
+	"ns-853.awsdns-42.net",
+}
+
+var testDNSData = DNSData{
+	DNSQuery{"A", "a." + testDomain}: DNSAnswers{
+		{"A", "2.2.2.2"},
+		{"A", "1.1.1.1"},
+	},
+
+	DNSQuery{"AAAA", "aaaa." + testDomain}: DNSAnswers{
+		{"AAAA", "2001:db8::aaaa"},
+	},
+
+	DNSQuery{"CNAME", "terratest." + testDomain}: DNSAnswers{
+		{"CNAME", "gruntwork-io.github.io."},
+	},
+
+	DNSQuery{"CNAME", "cname1." + testDomain}: DNSAnswers{
+		{"CNAME", "cname2." + testDomain + "."},
+	},
+
+	DNSQuery{"A", "cname1." + testDomain}: DNSAnswers{
+		{"CNAME", "cname2." + testDomain + "."},
+		{"CNAME", "cname3." + testDomain + "."},
+		{"CNAME", "cname4." + testDomain + "."},
+		{"CNAME", "cnamefinal." + testDomain + "."},
+		{"A", "1.1.1.1"},
+	},
+
+	DNSQuery{"TXT", "txt." + testDomain}: DNSAnswers{
+		{"TXT", `"This is a text."`},
+	},
+
+	DNSQuery{"MX", testDomain}: DNSAnswers{
+		{"MX", "10 mail." + testDomain + "."},
+	},
+}
+
+func TestOkDNSFindNameservers(t *testing.T) {
+	t.Parallel()
+	fqdn := "terratest.gruntwork.io"
+	expectedNameservers := publicDomainNameservers
+	nameservers, err := DNSFindNameserversE(t, fqdn, nil)
+	require.NoError(t, err)
+	require.ElementsMatch(t, nameservers, expectedNameservers)
+}
+
+func TestErrorDNSFindNameservers(t *testing.T) {
+	t.Parallel()
+	fqdn := "this.domain.doesnt.exist"
+	nameservers, err := DNSFindNameserversE(t, fqdn, nil)
+	require.Error(t, err)
+	require.Nil(t, nameservers)
+}
+
+func TestOkTerratestDNSLookupAuthoritative(t *testing.T) {
+	t.Parallel()
+	dnsQuery := DNSQuery{"CNAME", "terratest." + testDomain}
+	expected := DNSAnswers{{"CNAME", "gruntwork-io.github.io."}}
+	res, err := DNSLookupAuthoritativeE(t, dnsQuery, nil)
+	require.NoError(t, err)
+	require.ElementsMatch(t, res, expected)
+}
+
+func TestOkLocalDNSLookupAuthoritative(t *testing.T) {
+	t.Parallel()
+	ns1, ns2, dnsData1, dnsData2 := setupTestDNSServers()
+	dnsData1 = dnsData2
+	for dnsQuery, expected := range testDNSData {
+		(*dnsData1)[dnsQuery] = expected
```

Comment:
This is a bit confusing to me, especially since the name is not very descriptive. Two confusing elements:

- Why is it ok to update this on every loop? Does this pollute the next check?
- Why are we throwing away `dnsData1`? Consider doing the following, which is easier to follow:

```go
ns1, ns2, _, dnsData := setupTestDNSServers()
// use `dnsData` where using `dnsData1`
```

Reply:
Ohhh I think I get what's going on now. This is an internal database to handle the internal DNS data...

This is actually pretty confusing, so either:

- Update to a model where you pass in the database data to `setupTestDNSServers` as a function arg (or define a function `addEntryToDNSDatabase`).
- Add ample comments to explain what is going on.
- Rename `dnsData` to `dnsDatabase`.

Reply:
That was a poor attempt to reduce lines of code when both test servers must use the same answers, that in the end made things confusing.
Refactored the whole DNSData concept to embed internal DNSDatabase and DNSDatabaseRetry in each dnsTestServer.
New entries can be added using the dnsTestServer.addEntryToDNSDatabase method.